utils/user_management.py

```diff
@@ -0,0 +1,158 @@
+    # type
+    punctuation = '!#$%&()*+,-./:;<=>?@[\\]^_{|}~'
+
+    # we also want to have more letters and digits in
+    # generated password than punctuations
+    password_chars =\
+        lower_case_letters + lower_case_letters +\
+        upper_case_letters + upper_case_letters +\
+        digits + digits +\
+        punctuation
+    pwd = ''.join(random.choice(password_chars) for i in range(pwd_len))
+    logger.info('Generated password: "{}"'.format(pwd))
+    return pwd
+
+
+def store_passwords_file(pwdfile, credentials):
+    with open(pwdfile, 'wt') as f:
+        for username, pwd in credentials.items():
+            f.write('{} {}\n'.format(username, pwd))
+    return True
+
+
+def add_user(args, credentials):
+    username = args.username.lower()
+    logger.info('Adding username "{}"'.format(username))
+
+    if username in credentials:
+        logger.error('Can\'t add username {} as it is already present '
+                     'in passwords file. Do you want to run '
+                     '"update" command instead?'.format(username))
+        return False
+
+    password = args.password
+    logger.info('Adding username "{}" with password "{}"...'.format(
+                (username, password)))
+    credentials[username] = hash_password(username, password)
+
+    return store_passwords_file(args.pwdfile, credentials)
+
+
+def update_user(args, credentials):
+    username = args.username.lower()
+    logger.info('Updating username "{}"'.format(username))
+
+    if username not in credentials:
+        logger.error('Username "{}" not found in passwords file. '
+                     'Do you want to run "add" command instead?'.
+                     format(username))
+        return False
+
+    password = args.password
+    logger.info('Updating username "{}" password  to "{}"...'.format(
+                (username, password)))
+    credentials[username] = hash_password(username, password)
+    return store_passwords_file(args.pwdfile, credentials)
+
+
+def process_command(args, credentials):
+    if args.command == 'add':
+        return add_user(args, credentials)
+    elif args.command == 'update':
+        return update_user(args, credentials)
+    else:
+        logger.error('Uknown command "%s"' % args.command)
+        return False
+
+
+def main():
+    parser = build_cli_parser()
+    args = parser.parse_args()
+    if not check_args(args):
+        parser.print_help()
+        return
+
+    succeeded, credentials = parse_pwd_file(args.pwdfile)
+    if not succeeded and args.command != 'add':
+        return
+
+    if args.password is None:
+        args.password = generate_password()
+
+    process_command(args, credentials)
+    return
+
+
+if __name__ == '__main__':
+    logging.basicConfig(level=logging.DEBUG, format="%(message)s")
+
+    # add tabpy-tools and tabpy-server folders to
+    # PYTHONPATH so code from there can be found when
+    # modules are imported
+    for dir_ in {'tabpy-tools', 'tabpy-server'}:

```

Comment:
Is there a reason for this? Generally modifying the path like this from within code makes it generally more difficult to debug. 

At the very least I don't think you need to add the tools package here as you aren't importing from it

Reply:
The main reason we want for users to be able to run the tool from repo root folder and for that we need to have at least `tabpy-server` in `PYTHONPATH` as we reuse some code from it. As for `tabpy-tools` it is not needed now but may be needed in future - server code still uses some types from tools package :(

Reply:
Gotcha. Not super important for now since this is just a util file but hoping we can effectively decouple the tools and server packages. Otherwise we should just bundle them together in code base

Reply:
Oh, believe me, we are constantly discussing how we can setup process more robust and platform independent.
There is a list of options but neither of them is going to work smoothly for all of the scenarios people use TabPy with.